contracts/src/v2/Calls.sol

```diff
@@ -0,0 +1,222 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023 Snowfork <hello@snowfork.com>
+pragma solidity 0.8.25;
+
+import {IERC20} from "../interfaces/IERC20.sol";
+import {IGateway} from "../interfaces/IGateway.sol";
+import {WETH9} from "canonical-weth/WETH9.sol";
+
+import {SafeNativeTransfer, SafeTokenTransfer} from "../utils/SafeTransfer.sol";
+
+import {AssetsStorage, TokenInfo} from "../storage/AssetsStorage.sol";
+import {CoreStorage} from "../storage/CoreStorage.sol";
+import {PricingStorage} from "../storage/PricingStorage.sol";
+import {SubstrateTypes} from "../SubstrateTypes.sol";
+import {MultiAddress} from "../MultiAddress.sol";
+import {Address} from "../utils/Address.sol";
+import {AgentExecutor} from "../AgentExecutor.sol";
+import {Agent} from "../Agent.sol";
+import {Call} from "../utils/Call.sol";
+import {Token} from "../Token.sol";
+import {Upgrade} from "../Upgrade.sol";
+import {Functions} from "../Functions.sol";
+import {Constants} from "../Constants.sol";
+
+import {Ticket, OperatingMode} from "./Types.sol";
+
+import {UD60x18, ud60x18, convert} from "prb/math/src/UD60x18.sol";
+
+/// @title Library for implementing Ethereum->Polkadot ERC20 transfers.
+library CallsV2 {
+    using Address for address;
+    using SafeTokenTransfer for IERC20;
+    using SafeNativeTransfer for address payable;
+
+    error InvalidProof();
+    error InvalidNonce();
+    error NotEnoughGas();
+    error FeePaymentToLow();
+    error Unauthorized();
+    error Disabled();
+    error AgentAlreadyCreated();
+    error AgentDoesNotExist();
+    error ChannelAlreadyCreated();
+    error ChannelDoesNotExist();
+    error InvalidChannelUpdate();
+    error AgentExecutionFailed(bytes returndata);
+    error InvalidAgentExecutionPayload();
+    error InvalidConstructorParams();
+    error AlreadyInitialized();
+    error TokenNotRegistered();
+    error InvalidAsset();
+
+    address public constant WETH_ADDRESS = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+    uint8 public constant MAX_ASSETS = 8;
+
+    // Send an XCM with arbitrary assets to Polkadot Asset Hub
+    //
+    // Params:
+    //   * `xcm` (bytes): SCALE-encoded XCM message
+    //   * `assets` (bytes[]): Array of asset specs.
+    //
+    // Supported asset specs:
+    // * ERC20: abi.encode(0, tokenAddress, value)
```

Comment:
What is the 0 for?

Reply:
Its a tag used to descriminate between different assets kinds. 

Its very much like a Rust enum. Since Solidity does not have those, we have to emulate them.